cmd/algokey/keyreg.go

```diff
@@ -0,0 +1,266 @@
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package main
+
+import (
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/spf13/cobra"
+
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/data/account"
+	"github.com/algorand/go-algorand/data/basics"
+	"github.com/algorand/go-algorand/data/transactions"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/util"
+	"github.com/algorand/go-algorand/util/db"
+)
+
+var keyregCmd *cobra.Command
+
+type keyregCmdParams struct {
+	fee         uint64
+	firstValid  uint64
+	lastValid   uint64
+	network     string
+	offline     bool
+	txFile      string
+	partkeyFile string
+	addr        string
+}
+
+// There is no node to query, so we do our best here.
+const (
+	txnLife uint64 = 1000
+	minFee  uint64 = 1000
+)
+
+type networkGenesis struct {
+	id   string
+	hash crypto.Digest
+}
+
+var validNetworks map[string]networkGenesis
+var validNetworkList []string
+
+func init() {
+	var params keyregCmdParams
+
+	keyregCmd = &cobra.Command{
+		Use:   "keyreg",
+		Short: "Make key registration transaction",
+		Args:  cobra.NoArgs,
+		Run: func(cmd *cobra.Command, _ []string) {
+			err := run(params)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "%s\n\n", err.Error())
+				os.Exit(1)
+			}
+		},
+	}
+
+	keyregCmd.Flags().Uint64Var(&params.fee, "fee", minFee, "transaction fee")
+	keyregCmd.Flags().Uint64Var(&params.firstValid, "first-valid", 0, "first round where the transaction may be committed to the ledger")
+	keyregCmd.MarkFlagRequired("first-valid")
+	keyregCmd.Flags().Uint64Var(&params.lastValid, "last-valid", 0, fmt.Sprintf("last round where the generated transaction may be committed to the ledger, defaults to first-valid + %d", txnLife))
+	keyregCmd.Flags().StringVar(&params.network, "network", "mainnet", "the network where the provided keys will be registered, one of mainnet/testnet/betanet")
+	keyregCmd.MarkFlagRequired("network")
+	keyregCmd.Flags().BoolVar(&params.offline, "offline", false, "set to bring an account offline")
+	keyregCmd.Flags().StringVar(&params.txFile, "tx-file", "", fmt.Sprintf("write signed transaction to this file, or '%s' to write to stdout", stdoutFilenameValue))
+	keyregCmd.MarkFlagRequired("tx-file")
+	keyregCmd.Flags().StringVar(&params.partkeyFile, "partkey-file", "", "participation keys to register, file is opened to fetch metadata for the transaction, mutually exclusive with account")
+	keyregCmd.Flags().StringVar(&params.addr, "account", "", "account address to bring offline, mutually exclusive with partkey-file")
+
+	// TODO: move 'bundleGenesisInject' into something that can be imported here instead of using constants.
+	validNetworks = map[string]networkGenesis{
```

Comment:
instead of hard-coding this, do this dynamically similar to `.go-algorand/cmd/goal/node.go`

Reply:
That command is also hard coded:
```
validNetworks := map[string][]byte{
			"mainnet": genesisMainnet,
			"testnet": genesisTestnet,
			"betanet": genesisBetanet,
			"devnet":  genesisDevnet,
		}
```

The reason I didn't use those objects is that they aren't readily accessible and I didn't want to refactor anything as part of this PR.